mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala

```diff
@@ -281,17 +295,28 @@ class Word2Vec extends Serializable with Logging {
     val expTable = sc.broadcast(createExpTable())
     val bcVocab = sc.broadcast(vocab)
     val bcVocabHash = sc.broadcast(vocabHash)
-
-    val sentences: RDD[Array[Int]] = words.mapPartitions { iter =>
+    // each partition is a collection of sentences, will be translated into arrays of Index integer
+    val sentences: RDD[Array[Int]] = dataset.mapPartitions { sentenceIter =>
       new Iterator[Array[Int]] {
-        def hasNext: Boolean = iter.hasNext
+        var wordIter: Iterator[String] = null
+
+        def hasNext: Boolean = sentenceIter.hasNext || (wordIter != null && wordIter.hasNext)
 
         def next(): Array[Int] = {
           val sentence = ArrayBuilder.make[Int]
           var sentenceLength = 0
-          while (iter.hasNext && sentenceLength < MAX_SENTENCE_LENGTH) {
-            val word = bcVocabHash.value.get(iter.next())
-            word match {
+          // do translation of each word into its index in the vocabulary,
+          // do cutting only when the sentence is larger than maxSentenceLength
+          if ((wordIter == null || !wordIter.hasNext) && sentenceIter.hasNext) {
+            do {
```

Comment:
This is just for the edge case when the input contains empty sentences, the do while loop will skip empty sentences instead of generate an empty array as a result for the empty sentence.
of course, if the sentence splitter will skip empty sentences, this is not necessary, but that's out of this class's control. So, I just recommend to have the loop, it won't introduce much latency since condition check is so cheap and the loop will only run once for 99.999% cases.